### DISCUSSION TITLE
Makefile.Release(1143) : fatal error U1001: syntax error : illegal character '{'

### DISCUSSION THREAD

**Discussion by JaminShanti** [jamulussoftware/jamulus#1998] [Technical Support & Help]
Anyone know what I am missing here.  Acting like gcc is not defined.

Running, 
`.\windows\deploy_windows.ps1 -QtCompile64 'mingw81_64'`
Getting:
Makefile.Release(1143) : fatal error U1001: syntax error : illegal character '{' in macro
1143 line is,
`	${CC} -msse $(CFLAGS) $(INCPATH) -c ../libs/opus/celt/x86/x86cpu.c -o release/x86cpu.o`
CC is defined as, CC = gcc

My PC:
PS D:\projects\github\jamulus\build> which gcc
/c/ProgramData/chocolatey/bin/gcc
`PS D:\projects\github\jamulus\build> gcc --version
gcc.exe (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0`

Environment: 
install QT from choco install qt5-default C:\Qt\5.15.2\mingw81_64
.\vswhere.exe -property catalog_productDisplayVersion 15.9.38

**Comment by ann0see** [comment]
Not sure if anything else then MSVC works with the build script. I‘d just install the Visual Studio build tools and follow the compilation instructions in this repository.

**Comment by npostavs** [comment]
If you want to build with mingw, you need to use `qmake`, I think.

**Comment by JaminShanti** [comment]
I'll try that today thanks.
Because of the work I am in, I try to stick to Bourne and PowerShell command lines. Everything else is a luxury.

**Reply by JaminShanti** [reply]
I added qmake to my path, still errors.
`**********************************************************************
Using Visual Studio/Build Tools environment settings located at
D:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\build\vcvars64.bat
**********************************************************************

**********************************************************************
Using Qt binaries for Visual C++ located at
C:\Qt\5.15.2\mingw81_64\bin
**********************************************************************

**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.9.38
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'
Info: creating stash file D:\projects\github\jamulus\build\.qmake.stash
Project MESSAGE: building version "3.8.0dev-38252048" (intermediate in git repository)
Project MESSAGE: building version "3.8.0dev-38252048" (intermediate in git repository)
Project MESSAGE: building version "3.8.0dev-38252048" (intermediate in git repository)

Microsoft (R) Program Maintenance Utility Version 14.16.27045.0
Copyright (C) Microsoft Corporation.  All rights reserved.

        "D:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\nmake.exe" -f Makefile.Release

Microsoft (R) Program Maintenance Utility Version 14.16.27045.0
Copyright (C) Microsoft Corporation.  All rights reserved.

Makefile.Release(1143) : fatal error U1001: syntax error : illegal character '{' in macro
Stop.
NMAKE : fatal error U1077: '"D:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.
Native command nmake returned with exit code 2
At D:\projects\github\jamulus\windows\deploy_windows.ps1:37 char:9
+         Throw "Native command $Command returned with exit code $LastE ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Native command ...ith exit code 2:String) [], RuntimeException
    + FullyQualifiedErrorId : Native command nmake returned with exit code 2`

**Reply by npostavs** [reply]
I don't think the MSVC version of qmake is going to work with gcc. Your options are installing qmake + qt libs etc with [Msys2](https://www.msys2.org/), or else try Qt Creator. https://wiki.qt.io/MinGW

**Reply by JaminShanti** [reply]
I think this thread points out some limitations building on windows.  I'm not a C developer, maybe you would be able to walk around the pitfalls.

**Reply by ann0see** [reply]
Do you really need MinGW? Then you'd need to do the stuff by hand. The officially supported compiler for windows is MSVC.

**Reply by ann0see** [reply]
Also note that you'll need to install the qt multimedia plugin.

Also linking https://stackoverflow.com/questions/71458639/qt-application-cannot-run-in-release here - you might need it.

**Reply by dcorson-ticino-com** [reply]
What stuff would need to be done by hand using the ming compiler ?
I have done all my local development using QT Creator with the ming compiler which is supplied with it and have changed nothing in the tools.
I never saw any incompatibilities after compiling on git either.
Was never able to get the MSVC compiler to work locally.

**Reply by ann0see** [reply]
I'm currently trying to build using mingw on the command line and it compiles but doesn't run (due to missing DLLs. I'll probably need to fiddle around with windeployqt).

Edit: Now it works:

I downloaded mingw from here: https://winlibs.com/ 
and installed qt for mingw. Then I run` C:\path\to\qmake.exe Jamulus.pro` set the PATH to the compiler as environment variable, then run `mingw32-make.exe -j <numberOfProcesses>`

Then `C:\Qt\6.5.1\mingw_64\bin\qtenv2.bat` to set up the environment, 

and then `C:\Qt\6.5.1\mingw_64\bin\windeployqt.exe`

**Comment by pljones** [comment]
I'd suggest starting with an official Qt build for Windows with Qt Creator and all plugins, both latest Qt6 and latest Qt5 for all provided targets.

Then you'll be able to pick the Qt5 kit using the _provided_ GnuCC compiler - I've not yet tried this... it works with the MSVC kit, though.

Once you have builds working from Qt Creator, you can inspect what it's doing and try recreating this - or inspect what the build chains are doing on Github.  Me, I just want it built locally for testing.  For deployment, I'd use the Github version every time, as you then know it'll be what goes into the official release.

(If anyone gets Qt Creator builds on either Windows or Linux running targetting any Android version, I'd love to hear - not in this discussion, though.)
